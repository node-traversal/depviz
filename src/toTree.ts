import isUndefined from 'lodash/isUndefined';
import forIn from 'lodash/forIn';
import Dependency from "./Dependency";

type TreeOptionsToggles = Record<string, boolean>
type TreeData = Record<string, Dependency>
interface TreeOptions {
    noTraverse?: TreeOptionsToggles;
    hidden?: TreeOptionsToggles;
}

type DataItem = {
    deps: DataDep[];
}

type DataDep = {
    name: string;
}

export default (data: object, options: TreeOptions): Dependency[] => {
    const result: Dependency[] = [];
    const cache: TreeData = {};
    const noTraverse = options.noTraverse || {};
    const hidden = options.hidden || {};
    const shouldTraverse = (key: string): boolean => isUndefined(noTraverse[key]);
    const isVisible = (key: string): boolean => isUndefined(hidden[key]);

    forIn(data, (item: DataItem, key: string) => {
        const node: Dependency = cache[key] || new Dependency(key, []);
        const rawDeps: DataDep[] = item.deps;
        if (rawDeps && shouldTraverse(key)) {
            rawDeps.forEach((dep) => {
                if (isVisible(dep.name)) {
                    let prevDep = cache[dep.name];

                    if (!prevDep) {
                        prevDep = new Dependency(dep.name, []);
                    }

                    node.deps.push(prevDep);
                    cache[dep.name] = prevDep;
                }
            });
        }

        result.push(node);
        cache[key] = node;
    });

    return result;
};