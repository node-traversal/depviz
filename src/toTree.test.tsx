import toTree from './toTree';
import data from './sample-data.json';

it('convert empty data to tree', () => {
    expect(toTree({}, {})).toEqual([]);
});

it('convert omit hidden nodes', () => {
    expect(toTree(data, {
        hidden: {
            react: true
        }
    })).toMatchSnapshot();
});

it('converted nodes with noTraverse must not have dependencies', () => {
    expect(toTree(data, {
        noTraverse: {
            '@examples/simple': true
        }
    })).toMatchSnapshot();
});

it('convert data to tree', () => {
    expect(toTree(data, {})).toMatchSnapshot();
});
