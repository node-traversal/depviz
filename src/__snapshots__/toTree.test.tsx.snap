// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convert data to tree 1`] = `
Array [
  Dependency {
    "_deps": Array [
      Dependency {
        "_deps": Array [],
        "_name": "react",
      },
      Dependency {
        "_deps": Array [],
        "_name": "react-dom",
      },
      Dependency {
        "_deps": Array [
          Dependency {
            "_deps": Array [],
            "_name": "react",
          },
          Dependency {
            "_deps": Array [],
            "_name": "prop-types",
          },
        ],
        "_name": "@examples/simple/simple-example",
      },
    ],
    "_name": "@examples/simple",
  },
  Dependency {
    "_deps": Array [
      Dependency {
        "_deps": Array [],
        "_name": "react",
      },
      Dependency {
        "_deps": Array [],
        "_name": "prop-types",
      },
    ],
    "_name": "@examples/simple/simple-example",
  },
]
`;

exports[`convert omit hidden nodes 1`] = `
Array [
  Dependency {
    "_deps": Array [
      Dependency {
        "_deps": Array [],
        "_name": "react-dom",
      },
      Dependency {
        "_deps": Array [
          Dependency {
            "_deps": Array [],
            "_name": "prop-types",
          },
        ],
        "_name": "@examples/simple/simple-example",
      },
    ],
    "_name": "@examples/simple",
  },
  Dependency {
    "_deps": Array [
      Dependency {
        "_deps": Array [],
        "_name": "prop-types",
      },
    ],
    "_name": "@examples/simple/simple-example",
  },
]
`;

exports[`converted nodes with noTraverse must not have dependencies 1`] = `
Array [
  Dependency {
    "_deps": Array [],
    "_name": "@examples/simple",
  },
  Dependency {
    "_deps": Array [
      Dependency {
        "_deps": Array [],
        "_name": "react",
      },
      Dependency {
        "_deps": Array [],
        "_name": "prop-types",
      },
    ],
    "_name": "@examples/simple/simple-example",
  },
]
`;
